controller/artifact.go

```diff
@@ -31,33 +35,49 @@ func (r *ArtifactRepo) Add(data interface{}) error {
 	if a.URI == "" {
 		return ct.ValidationError{Field: "uri", Message: "must not be empty"}
 	}
+	if a.Type == ct.ArtifactTypeFlynn && a.Manifest == nil {
+		if a.Size <= 0 {
+			return ct.ValidationError{Field: "size", Message: "must be greater than zero"}
+		}
+		if err := downloadManifest(a); err != nil {
+			return ct.ValidationError{Field: "manifest", Message: fmt.Sprintf("failed to download from %s: %s", a.URI, err)}
+		}
+	}
+
 	tx, err := r.db.Begin()
 	if err != nil {
 		return err
 	}
 
-	err = tx.QueryRow("artifact_insert", a.ID, string(a.Type), a.URI, a.Meta).Scan(&a.CreatedAt)
+	err = tx.QueryRow("artifact_insert", a.ID, string(a.Type), a.URI, a.Meta, a.Manifest, a.Hashes, a.Size, a.LayerURLTemplate).Scan(&a.CreatedAt)
```

Comment:
Should we persist the raw JSON manifest instead of the round-tripped manifest so that the hashes continue to match?